File: 2024/d12/main.py
input = open(0).read()


g = input.splitlines()
rows = len(g)
cols = len(g[0])


dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]
seen = set()


def dfs(r: int, c: int, f: dict[str, int]):
    seen.add((r, c))
    f["count"] += 1

    def is_neighbour(dr, dc):
        nr, nc = r + dr, c + dc
        return 0 <= nr < rows and 0 <= nc < cols and g[nr][nc] == g[r][c]

    for dr, dc in dirs:
        nr, nc = r + dr, c + dc
        # if not (0 <= nr < rows and 0 <= nc < cols) or g[r][c] != g[nr][nc]:
        if not is_neighbour(dr, dc):
            f["sides"] += 1
        elif (nr, nc) not in seen:
            dfs(nr, nc, f)


def dfs2(r: int, c: int, f: dict[str, int]):
    seen.add((r, c))
    f["count"] += 1

    def is_neighbour(dr, dc):
        nr, nc = r + dr, c + dc
        return 0 <= nr < rows and 0 <= nc < cols and g[nr][nc] == g[r][c]

    for d in range(4):
        dr, dc = dirs[d]
        ndr, ndc = dirs[(d + 1) % 4]

        if not is_neighbour(dr, dc) and not is_neighbour(ndr, ndc):
            f["sides"] += 1

        if (
            is_neighbour(dr, dc)
            and is_neighbour(ndr, ndc)
            and not is_neighbour(dr + ndr, dc + ndc)
        ):
            f["sides"] += 1

    for dr, dc in dirs:
        nr, nc = r + dr, c + dc
        if is_neighbour(dr, dc) and (nr, nc) not in seen:
            dfs2(nr, nc, f)


out = 0
for r in range(rows):
    for c in range(cols):
        if (r, c) not in seen:
            fence = dict(count=0, sides=0)
            # dfs(r, c, fence)
            dfs2(r, c, fence)
            out += fence["count"] * fence["sides"]
            print((r, c), fence)

print(out)

File: 2024/d12/main.cpp
#include <bits/stdc++.h>

using namespace std;

using pos = pair<int, int>;

vector<string> g;
int rows, cols;
set<pos> seen;

pos dirs[4] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

bool inbounds(int r, int c) {
  return 0 <= r && r < rows && 0 <= c && c < cols;
}

void dfs(int r, int c, int& count, int& sides) {
  seen.insert({r, c});
  count++;
  for (pos dir : dirs) {
    int nr = r + dir.first, nc = c + dir.second;
    if (!inbounds(nr, nc) || g[nr][nc] != g[r][c]) {
      sides++;
    } else if (seen.count({nr, nc}) == 0) {
      dfs(nr, nc, count, sides);
    }
  }
}

void dfs2(int r, int c, int& count, int& sides) {
  seen.insert({r, c});
  count++;

  auto isneighbour = [&](pos dir) -> bool {
    int nr = r + dir.first, nc = c + dir.second;
    return inbounds(nr, nc) && g[nr][nc] == g[r][c];
  };

  for (int d = 0; d < 4; d++) {
    pos dir1 = dirs[d];
    pos dir2 = dirs[(d + 1) % 4];
    if (!isneighbour(dir1) && !isneighbour(dir2)) {
      sides++;
    }
    if (isneighbour(dir1) && isneighbour(dir2) &&
        !isneighbour({dir1.first + dir2.first, dir1.second + dir2.second})) {
      sides++;
    }
  }

  for (pos dir : dirs) {
    int nr = r + dir.first, nc = c + dir.second;
    if (isneighbour(dir) && seen.count({nr, nc}) == 0) {
      dfs2(nr, nc, count, sides);
    }
  }
}

int main() {
  for (string line; getline(cin, line);) g.push_back(line);
  rows = g.size();
  cols = g[0].size();

  long out = 0;
  for (int r = 0; r < rows; r++) {
    for (int c = 0; c < cols; c++) {
      if (seen.count({r, c}) != 0) continue;
      int count = 0, sides = 0;
      dfs2(r, c, count, sides);
      printf("start=(%d,%d) count=%d sides=%d\n", r, c, count, sides);
      out += count * sides;
    }
  }

  cout << out << endl;
}
